src/utils/src/ThreadsafeBlockingQueue.c

```diff
@@ -0,0 +1,214 @@
+    CHK_STATUS(stackQueueCreate(&(pSafeQueue->queue)));
+
+    *ppSafeQueue = pSafeQueue;
+
+CleanUp:
+    if (STATUS_FAILED(retStatus) && pSafeQueue != NULL) {
+        SAFE_MEMFREE(pSafeQueue);
+    }
+
+    return retStatus;
+}
+
+/**
+ * Frees and de-allocates the thread safe blocking queue
+ */
+STATUS safeBlockingQueueFree(PSafeBlockingQueue pSafeQueue)
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    CHK(pSafeQueue != NULL, STATUS_NULL_ARG);
+
+    // set terminate flag, lock mutex -- this assures all other threads
+    // are no longer directly interacting with the queue
+    //
+    // free semaphore, this will unblock any threads blocking on said semaphore,
+    // and then exit from the terminate flag
+    //
+    // unlock mutex
+    ATOMIC_STORE_BOOL(&pSafeQueue->terminate, TRUE);
+
+    MUTEX_LOCK(pSafeQueue->mutex);
+    CHK_STATUS(semaphoreFree(&(pSafeQueue->semaphore)));
+    MUTEX_UNLOCK(pSafeQueue->mutex);
+
+    CHK_STATUS(stackQueueFree(pSafeQueue->queue));
+    MUTEX_FREE(pSafeQueue->mutex);
+
+    SAFE_MEMFREE(pSafeQueue);
+
+CleanUp:
+
+    return retStatus;
+}
+
+/**
+ * Clears and de-allocates all the items
+ */
+STATUS safeBlockingQueueClear(PSafeBlockingQueue pSafeQueue, BOOL freeData)
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    BOOL locked = FALSE;
+
+    CHK(pSafeQueue != NULL, STATUS_NULL_ARG);
+    CHK(!ATOMIC_LOAD_BOOL(&pSafeQueue->terminate), STATUS_INVALID_OPERATION);
+
+    // 0 timeout semaphore acquire, eventually getting a timeout, this is done to clear
+    // the counting semaphore
+    while ((semaphoreAcquire(pSafeQueue->semaphore, 0) == STATUS_SUCCESS) && !ATOMIC_LOAD_BOOL(&pSafeQueue->terminate))
+        ;
+
+    CHK(!ATOMIC_LOAD_BOOL(&pSafeQueue->terminate), STATUS_INVALID_OPERATION);
+
+    MUTEX_LOCK(pSafeQueue->mutex);
+    locked = TRUE;
+    CHK(!ATOMIC_LOAD_BOOL(&pSafeQueue->terminate), STATUS_INVALID_OPERATION);
+
+    CHK_STATUS(stackQueueClear(pSafeQueue->queue, freeData));
+
+    MUTEX_UNLOCK(pSafeQueue->mutex);
+    locked = FALSE;
+
+CleanUp:
+    if (locked) {
+        MUTEX_UNLOCK(pSafeQueue->mutex);
+    }
+
+    return retStatus;
+}
+
+/**
+ * Gets the number of items in the stack/queue
+ */
+STATUS safeBlockingQueueGetCount(PSafeBlockingQueue pSafeQueue, PUINT32 pCount)
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    BOOL locked = FALSE;
+
+    CHK(pSafeQueue != NULL, STATUS_NULL_ARG);
+    CHK(!ATOMIC_LOAD_BOOL(&pSafeQueue->terminate), STATUS_INVALID_OPERATION);
+
+    MUTEX_LOCK(pSafeQueue->mutex);
+    locked = TRUE;
+    CHK(!ATOMIC_LOAD_BOOL(&pSafeQueue->terminate), STATUS_INVALID_OPERATION);
+
+    CHK_STATUS(stackQueueGetCount(pSafeQueue->queue, pCount));
+
+    MUTEX_UNLOCK(pSafeQueue->mutex);
+    locked = FALSE;
+
+CleanUp:
+    if (locked) {
+        MUTEX_UNLOCK(pSafeQueue->mutex);
+    }
+
+    return retStatus;
+}
+
+/**
+ * Whether the thread safe blocking queue is empty
+ */
+STATUS safeBlockingQueueIsEmpty(PSafeBlockingQueue pSafeQueue, PBOOL pIsEmpty)
```

Comment:
Implement this as a call to get count and return whether the count is 0

Reply:
The implementation for ThreadsafeBlockingQueue directly maps onto StackQueue. There is an operation for count in StackQueue, but I won't be needing it for what I have planned for this queue, nor do any future planned uses.

If a need comes up we can add it then.